src/red_black_tree.rs

```diff
@@ -331,235 +443,257 @@ impl<
-                    }
-                    self.transplant(ref_node_index, min_right);
-                    self.connect(min_right, left, Field::Left as u32);
-                    if self.is_red(ref_node_index) {
-                        self.color_red(min_right)
-                    } else {
-                        self.color_black(min_right)
-                    }
-                    self.allocator
-                        .clear_register(ref_node_index, Field::Left as u32);
-                    self.allocator
-                        .clear_register(ref_node_index, Field::Right as u32);
-                    (min_right_child, ref_node_index)
-                };
-                self.allocator
-                    .clear_register(ref_node_index, Field::Parent as u32);
-                self.allocator.clear_register(delete_node_index, COLOR);
-                self.allocator.remove_node(delete_node_index);
-                if is_black && self.fix_remove(pivot_node_index) == None {
-                    return None;
+            let RBNode {
+                key: curr_key,
+                value: curr_value,
+            } = *self.allocator.get(curr_node_index).get_value();
+            let target = match key.cmp(&curr_key) {
+                Ordering::Less => self.get_left(curr_node_index),
+                Ordering::Greater => self.get_right(curr_node_index),
+                Ordering::Equal => {
+                    self._remove_tree_node(curr_node_index);
+                    return Some(curr_value);
                 }
-                return Some(ref_value);
             };
             if target == SENTINEL {
                 return None;
             }
-            ref_node_index = target
+            curr_node_index = target
+        }
+    }
+
+    fn _remove_tree_node(&mut self, node_index: u32) {
+        let mut is_black = self.is_black(node_index);
+        let left = self.get_left(node_index);
+        let right = self.get_right(node_index);
+        let (pivot_node_index, parent_and_dir) = if self.is_leaf(node_index) {
+            if !self.is_root(node_index) {
+                let parent = self.get_parent(node_index);
+                let dir = self._child_dir(parent, node_index);
+                // Remove pointer to the removed leaf node
+                self._connect(parent, SENTINEL, dir);
+                (SENTINEL, Some((parent, dir)))
+            } else {
+                // Set the root to SENTINEL
+                self.root = SENTINEL;
+                (SENTINEL, None)
+            }
+        } else if left == SENTINEL {
+            self._transplant(node_index, right);
+            (right, None)
+        } else if right == SENTINEL {
+            self._transplant(node_index, left);
+            (left, None)
+        } else {
+            // Find the largest node in the left subtree
+            let mut parent_and_dir = None;
+            let max_left = self._find_max(left);
```

Comment:
I decided to use max(left tree) instead of min(right tree) as this the algorithm utilized by this visualizer: https://www.cs.usfca.edu/~galles/visualization/RedBlack.html

Was able to follow the exact topology changes without any ambiguity